cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,960 @@
+			},
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			unequalParts: []string{"inboundRules"},
+			equal:        false,
+		},
+		{
+			name: "detects when outboundRules are not equal",
+			firewall: &godo.Firewall{
+				ID:            "123",
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testDiffOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			unequalParts: []string{"outboundRules"},
+			equal:        false,
+		},
+		{
+			name: "detects when tags are not equal",
+			firewall: &godo.Firewall{
+				ID:            "123",
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          []string{"rick", "and", "morty"},
+			},
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			unequalParts: []string{"tags"},
+			equal:        false,
+		},
+		{
+			name: "detects when more then one property is not equal",
+			firewall: &godo.Firewall{
+				ID:            "123",
+				Name:          testWorkerFWName,
+				InboundRules:  testDiffInboundRules,
+				OutboundRules: testDiffOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			unequalParts: []string{"inboundRules", "outboundRules"},
+			equal:        false,
+		},
+		{
+			name: "detects when all properties are not equal",
+			firewall: &godo.Firewall{
+				Name:          "different name",
+				InboundRules:  testDiffInboundRules,
+				OutboundRules: testDiffOutboundRules,
+				Tags:          []string{"rick", "and", "morty"},
+			},
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			unequalParts: []string{"name", "inboundRules", "outboundRules", "tags"},
+			equal:        false,
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			equal, unequalParts := isEqual(test.firewall, test.firewallRequest)
+			if test.equal != equal {
+				t.Errorf("isEqual() mismatch (want: %v, got: %v)", test.equal, equal)
+			}
+			if diff := cmp.Diff(test.unequalParts, unequalParts); diff != "" {
+				t.Errorf("isEqual() mismatch (-want +got):\n%s", diff)
```

Comment:
This failure message is almost identical to the one in line 703. Could we include that this one is about the equal parts and the one above about the equal boolean value?